Verify configured voting and signing keys when checking validator status

sisou opened the pull request:
To prevent validator nodes running undetected with wrong configured voting and signing keys, I added checks to the inactive check to compare the configured keys to the on-chain keys to alert node operators about a potential missmatch.

sisou committed:
Verify configured voting and signing keys when checking validator status

sisou commented:
@viquezclaudio Testing this, and looking at the code, it looks like `get_staking_state()` is called for every block (when the Validator is polled). Is that fine? Should I make the config check happen less often? In the usual case, once per node startup is enough (although the keys registered in the staking contract can be changed at any time, of course).

viquezclaudio commented:
Yes, I think it will be better to do it just once, at init

sisou commented:
@viquezclaudio I have moved the check to the `init_epoch()` method which gets called both on init and for each new epoch. That way the check is only done at init and every 12 hours, so it still re-checks if the validator was updated in the staking contract.

jsdanielh committed:
blockchain: Optimize the chain store block table

Optimize the chain store block table by differentiating pushed vs
stored blocks in the chain store.
Pushed blocks are partially restored and are rebuilt by using the
chain info DB and the history store. Stored blocks are fully stored
since there is no associated history.
The `chain_info.on_main_chain` flag is used to differentiate where
to store the block and how to rebuild it (if needed).

Eligioo commented on validator/src/validator.rs:
```diff
@@ -299,6 +300,20 @@ where
         });
+
+        // Check validator configuration
+        let _ = self.get_validator(&blockchain)
```
```suggestion
        if let Ok(Some(validator)) = self.get_validator(&blockchain) {
            ...
        }
```

As you are only interested in the `if let Ok(Some(validator))` case, this is basically the same without the nested `map`s.

Eligioo commented on validator/src/validator.rs:
```diff
@@ -630,25 +645,31 @@ where
     }
 
-    fn get_staking_state(&self, blockchain: &Blockchain) -> ValidatorStakingState {
+    fn get_validator(&self, blockchain: &Blockchain) -> Result<Option<ValidatorAccount>, ()> {
```
Why is the return type `Result<Option<ValidatorAccount>, ()>`? Returning `Option<ValidatorAccount>` already is enough and lets you get rid of the `()` error case.

Then `get_staking_state` could return `ValidatorStakingState::Unknown` in the case of `get_validator` returning `None`.

sisou replied:
Nope. Unknown is when we fail to read the staking contract. When it's none, it returns Inactive.

Eligioo replied:
Now if `get_validator` returns `Err(())`, it means that the `staking_contract` can't be found. That feels weird. You can only know this by reading the function. Result should then return a proper Error here instead of `()`.

Looking at it again you can even get rid of `ValidatorStakingState::NoStake` as it's a dead variant. Then it can be `get_validator -> Option<ValidatorAccount>`

sisou committed:
Simplify get_validator method

Eligioo reviewed (approved):
LGTM

jsdanielh committed:
pow-migration: Do not use `Policy` functions to calculate epoch numbers

Change the blockchain history store interface to add a function
(`add_to_history_for_epoch`) that receives the epoch number instead
of relying on `Policy` to calculate it. This is useful in the history
migration for `PoW` since the genesis block number wouldn't be known
until the migration is over.
